### PR TITLE
[Editorial] Flight Route Requests

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [ ] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Comment by brebenelmihnea** [comment]
https://cses.fi/problemset/result/10007131/
image

**Review by envyaims — COMMENTED** [review]
wait so the only thing you used dsu for is to find all connected components and process them after?

File: solutions/gold/cses-1699.mdx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	int ans = 0;
+	for (int i = 0; i < components; i++) {
+		int sz = (int)t[i].size();
+		bool has_cycle = false;
+		for (int node : t[i]) {
+			if (!has_cycle && seen[node] == 0) { has_cycle |= dfs(node); }
+		}
+		ans += (has_cycle ? sz : sz - 1);
+	}
+	cout << ans;
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1699.mdx:108]:
```suggestion
	cout << ans << endl;
```

**Comment by ryanchou-dev** [comment]
> wait so the only thing you used dsu for is to find all connected components and process them after?

^?

File: solutions/gold/cses-1699.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply solutions/gold/cses-1699.mdx]: Done

**Comment by brebenelmihnea** [comment]
@envyaims yes. If you consider it's banal I'll remove it.

**Comment by SansPapyrus683** [comment]
yeah just do a dfs

**Comment by brebenelmihnea** [comment]
Yeah but like a DFS won't necessarily detect cycles on a DAG.

**Comment by envyaims** [comment]
> I don't think it would work, like a DFS won't necessarily detect cycles on a directed graph's connected component. That's why DSU is needed.

how come?

**Comment by brebenelmihnea** [comment]
> > I don't think it would work, like a DFS won't necessarily detect cycles on a directed graph's connected component. That's why DSU is needed.
> 
> how come? Arent you just using a DFS to get cycles in ur code?

Yes but at the same time I need to know the number of nodes in the connected component which is a lot more harder to computed with plain DFS in a directed graph. Let's say it's something like this:

image

DSU makes it easier.

**Comment by envyaims** [comment]
construct another undirected graph?

**Comment by brebenelmihnea** [comment]
I'm not sure what you mean by that.

**Comment by envyaims** [comment]
> I'm not sure what you mean by that.

I'm confused. Your DSU merges by path compression so it treats a and b are undirected edges. Idk what its doing that a CC on a regular graph can't do.

**Comment by brebenelmihnea** [comment]
Ah ok I see what you mean. In this case. Consider this:
image

The SCC on the undirected graph will determine in both cases there are cycles. Which is not true.

In the first case, we only need 2 edges to satisfy the reachability constraints, and in the second one we only need to add one more edge to satisfy the reachability constraints.

**Comment by envyaims** [comment]
> Ah ok I see what you mean. In this case. Consider this: image
> 
> The SCC on the undirected graph will determine in both cases there are cycles. Which is not true.
> 
> In the first case, we only need 2 edges to satisfy the reachability constraints, and in the second one we only need to add one more edge, resulting in 4 edges, to satisfy the reachability constraints.
> 
> The SCC thinking that there are cycles in both cases will think there should be 3 edges in the first case.

AC code without DSU
https://cses.fi/paste/923b1ee3daf801a298ce50/

**Comment by brebenelmihnea** [comment]
> construct another undirected graph?

Now I get it :)) I don't know what was I thinking.

Ok then, if you wish change my code with yours.

**Comment by envyaims** [comment]
> Ok then, if you wish change my code with yours. Or I'll just drop the pr.

move it to silver dfs or smth,

**Comment by SansPapyrus683** [comment]
wait mihnea...
i thought i said to ask me before adding a new problem

**Comment by brebenelmihnea** [comment]
> wait mihnea... i thought i said to ask me before adding a new problem
I can't recall that, but I'll do so form now on.

**Pull request closed without merging**